### PR TITLE
Implement Refresh Token

### LINKED ISSUE #160: Introduce Refresh Token for authentication
**What would you like to be added**:
Currently, the system only uses an Access Token for authentication. It is recommended to consider introducing a Refresh Token as an additional authentication mechanism.

**Why is this needed**:
The introduction of a Refresh Token will enhance the security and usability of the authentication process. It will allow users to obtain a new Access Token without having to reauthenticate every time it expires. This provides a seamless user experience and reduces the risk of unauthorized access due to expired tokens.

**Additional Information**:
Further considerations for the implementation of the Refresh Token mechanism should include:
- How long the Refresh Token should be valid
- The process or API endpoints for obtaining a new Access Token using a Refresh Token
- How to securely store and manage Refresh Tokens on the server-side
- Any potential impact on existing authentication flows and client applications

### PR DESCRIPTION
**Pull request by xet-a**
#### What this PR does / why we need it?
This PR implements a Refresh Token mechanism alongside the existing Access Token authentication method. The addition of a Refresh Token enhances security and usability by allowing users to seamlessly obtain new Access Tokens without the need for reauthentication every time the Access Token expires. 

The changes include:

- Frontend:
  - Saving the Refresh Token alongside the Access Token upon successful authentication.
  - Adding an Axios interceptor to automatically call the `/auth/refresh` endpoint when the Access Token is detected as expired.
  - Implementing a `POST` request to the `/auth/refresh` endpoint to retrieve a new Access Token.

- Backend:
  - Generating the Refresh Token upon user authentication.
  - Implementing a `RefreshStrategy` to manage Refresh Token validity and secure handling.
  - Creating a `POST /auth/refresh` API to generate a new Access Token using a valid Refresh Token.

Currently, the Access Token has a validity of **one day**, while the Refresh Token is valid for **one week**.

#### Any background context you want to provide?
The introduction of a Refresh Token aligns with best practices in API authentication and allows for more robust user sessions while minimizing the friction caused by token expiration. Careful consideration has been given to token lifecycle management and the implications for existing client applications and authentication flows.

#### What are the relevant tickets?

Fixes #160

### Checklist
- [x] Added relevant tests or not required
- [x] Didn't break anything

## Summary by CodeRabbit

- **New Features**
	- Introduced support for refresh tokens to enhance user session management.
	- Implemented token refresh logic in the application to improve authentication flow.
	- Enhanced JWT configuration for better security and management of access and refresh tokens.

- **Bug Fixes**
	- Improved error handling for user data retrieval and authentication states.

- **Style**
	- Reorganized import statements for clarity across several components.

- **Chores**
	- Updated state management in the Redux store to accommodate refresh tokens.

### CODE REVIEW

**Comment by xet-a** [comment]
I think there are still a few tasks left.
It works, but there is a problem that the error appears first and is processed.

image
First, I get a 401 error as shown in the picture when I first access. It's automatically redirected when the refresh token expires. I think it's the same as the second question, is there any other reason?

image
`const res = await axios.get<GetUserResponse>("/users");`
Second, when the access token expires, the 401 error comes out first, as shown in the photo, which seems to be caused by the way React Query works and the order of execution of the axios interceptor. React Query basically tries to detect and process the error immediately, which is likely to happen before the interceptor works.

Can you let me know if there is a good solution?

**Comment by devleejb** [comment]
@xet-a 
> First, I get a 401 error as shown in the picture when I first access. It's automatically redirected when the refresh token expires. I think it's the same as the second question, is there any other reason?

I this it is because of https://github.com/yorkie-team/codepair/issues/193.
After resolving the issue, I will merge it!

File: backend/src/auth/auth.service.ts [hunk 1]
```diff
@@ -1,7 +1,45 @@
...
+
+		const accessToken = this.jwtService.sign(
+			{ sub: user.id, nickname: user.nickname },
+			{ expiresIn: `${this.configService.get("JWT_ACCESS_TOKEN_EXPIRATION_TIME")}s` }
+		);
+
+		const refreshToken = this.jwtService.sign(
+			{ sub: user.id },
+			{ expiresIn: `${this.configService.get("JWT_REFRESH_TOKEN_EXPIRATION_TIME")}s` }
+		);
```
Comment (devleejb) [inline backend/src/auth/auth.service.ts:30]: It seems that it uses same jwt secret key.
We should use different jwt secret key.

File: backend/src/auth/types/login-response.type.ts [hunk 2]
```diff
@@ -3,4 +3,5 @@ import { ApiProperty } from "@nestjs/swagger";
 export class LoginResponse {
 	@ApiProperty({ type: String, description: "Access token for CodePair" })
 	accessToken: string;
+	refreshToken: string;
```
[the remark this answers is not in the document]
Comment (devleejb) [reply backend/src/auth/types/login-response.type.ts:6]: Please add swagger docs decorator.

File: frontend/src/App.tsx [hunk 3]
```diff
@@ -95,6 +99,47 @@ function App() {
...
+			dispatch(setUserData(null));
+			// axios.defaults.headers.common["Authorization"] = "";
+		};
+
+		const interceptor = axios.interceptors.response.use(
+			(response) => response,
+			async (error) => {
+				if (error.response?.status === 401) {
+					if (error.config.url === "/auth/refresh") {
+						handleRefreshTokenExpiration();
```
Comment (devleejb) [inline frontend/src/App.tsx:115]: It seems that we call this function twice.

```
              API Call 1 (401)        ->         `/auth/refresh` (Error)
                                                           ↓
`handleRefreshTokenExpiration()`      <-       `handleRefreshTokenExpiration()`
```

File: frontend/src/store/authSlice.ts [hunk 4]
```diff
@@ -17,10 +19,13 @@ export const authSlice = createSlice({
 		setAccessToken: (state, action: PayloadAction<string | null>) => {
 			state.accessToken = action.payload;
 		},
+		setRefreshToken(state, action: PayloadAction<string | null>) {
+			state.refreshToken = action.payload;
+		},
 	},
```
Comment (devleejb) [inline frontend/src/store/authSlice.ts:25]: How about add `logout` function here?

In many other files, we create logout logic.

**Review by devleejb — CHANGES_REQUESTED** [review]
It seems to be well implemented overall. I think it just needs a few details to be adjusted.

**Review by devleejb — CHANGES_REQUESTED** [review]
Also, I recommend to create two ConfigService objects like below code.
(Maybe, there are more good approaches.)

#### `auth.module.ts`
```typescript
@Module({
	imports: [UsersModule],
	providers: [
		AuthService,
		GithubStrategy,
		JwtStrategy,
		JwtRefreshStrategy,
		{
			provide: "JWT_ACCESS_SERVICE",
			useFactory: (configService: ConfigService) => {
				return new JwtService({
					secret: configService.get<string>("JWT_ACCESS_TOKEN_SECRET"),
					signOptions: {
						expiresIn: `${configService.get("JWT_ACCESS_TOKEN_EXPIRATION_TIME")}s`,
					},
				});
			},
			inject: [ConfigService],
		},
		{
			provide: "JWT_REFRESH_SERVICE",
			useFactory: (configService: ConfigService) => {
				return new JwtService({
					secret: configService.get<string>("JWT_REFRESH_TOKEN_SECRET"),
					signOptions: {
						expiresIn: `${configService.get("JWT_REFRESH_TOKEN_EXPIRATION_TIME")}s`,
					},
				});
			},
			inject: [ConfigService],
		},
	],
	exports: ["JWT_ACCESS_SERVICE", "JWT_REFRESH_SERVICE"],
	controllers: [AuthController],
})
export class AuthModule {}
```

#### `auth.service.ts`
```typescript
constructor(
	private readonly usersService: UsersService,
	private readonly jwtService: JwtService,
	private readonly configService: ConfigService,
	@Inject("JWT_ACCESS_SERVICE") private readonly jwtAccessService: JwtService,
	@Inject("JWT_REFRESH_SERVICE") private readonly jwtRefreshService: JwtService
) {
        // Check whether the tokens are intended value.
	console.log("JWT_ACCESS_SERVICE", jwtAccessService);
	console.log("JWT_REFRESH_SERVICE", jwtRefreshService);
}
```

Also, you can create the object name constants :)
For example, `@Inject(JwtInject.ACCESS)`.

File: frontend/src/App.tsx [hunk 5]
```diff
@@ -108,24 +108,18 @@ function App() {
...
+					} else {
 						error.config._retry = true;
 						const refreshToken = store.getState().auth.refreshToken;
-						try {
-							const response = await axios.post("/auth/refresh", { refreshToken });
-							const newAccessToken = response.data.accessToken;
-							dispatch(setAccessToken(newAccessToken));
-							axios.defaults.headers.common["Authorization"] =
-								`Bearer ${newAccessToken}`;
-							error.config.headers["Authorization"] = `Bearer ${newAccessToken}`;
```
Comment (devleejb) [inline frontend/src/App.tsx]: We should add this line.
In `error.config`, there's the old access token.

**Review by devleejb — CHANGES_REQUESTED** [review]

**Review by devleejb — APPROVED** [review]
Thank you for your contribution.

**Pull request merged**